File: scripts/fix_namespace.pl
#!/usr/bin/perl

#
# fix_namespace.pl - Universal namespace migration tool
#
# This script detects the current namespace(s) used in the project and converts them
# to new namespaces specified by the user.
#
# Usage: perl scripts/fix_namespace.pl [new_ns1] [new_ns2]
#
# Example: perl scripts/fix_namespace.pl yourname yourproject
#          (changes david303ttl::synthtemplate to yourname::yourproject)
#

use File::Find;
use File::Basename;

# Function to detect current namespaces from source files
sub detect_current_namespaces
{
    my @files_to_check = (
        'src/configuration.h',
        'src/engine/engine.h',
        'src/engine/patch.h',
        'src/engine/voice.h'
    );

    foreach my $file (@files_to_check) {
        if (-e $file) {
            open(FH, "<$file") or next;
            my @ns_stack = ();
            while (my $line = <FH>) {
                # Look for namespace declarations like: namespace foo {
                if ($line =~ /namespace\s+(\w+)\s*\{/) {
                    push @ns_stack, $1;
                    # If we have 2 namespaces, we found our pattern
                    if (@ns_stack >= 2) {
                        close(FH);
                        return ($ns_stack[0], $ns_stack[1]);
                    }
                }
                # Look for namespace reference like foo::bar::
                if ($line =~ /(\w+)::(\w+)::/) {
                    close(FH);
                    return ($1, $2);
                }
            }
            close(FH);
        }
    }
    return (undef, undef);  # Not found
}

# Detect current namespaces by scanning source files
($old_ns1, $old_ns2) = detect_current_namespaces();

if ($old_ns1 && $old_ns2) {
    print "Detected current namespaces: '$old_ns1' :: '$old_ns2'\n";
} else {
    print "Could not detect current namespaces.\n";
    print "Enter first namespace (e.g., yourname): ";
    $old_ns1 = <STDIN>;
    chomp $old_ns1;
    print "Enter second namespace (e.g., yourproject): ";
    $old_ns2 = <STDIN>;
    chomp $old_ns2;
}

# Get new namespaces from command line arguments, or prompt for them
if (@ARGV >= 2) {
    $new_ns1 = $ARGV[0];
    $new_ns2 = $ARGV[1];
} elsif (@ARGV == 1) {
    $new_ns1 = $ARGV[0];
    print "Enter second new namespace (current: '$old_ns2', press Enter to keep): ";
    $input = <STDIN>;
    chomp $input;
    $new_ns2 = $input || $old_ns2;  # Keep old if Enter pressed
} else {
    print "Enter first new namespace (current: '$old_ns1'): ";
    $new_ns1 = <STDIN>;
    chomp $new_ns1;
    print "Enter second new namespace (current: '$old_ns2', press Enter to keep): ";
    $input = <STDIN>;
    chomp $input;
    $new_ns2 = $input || $old_ns2;  # Keep old if Enter pressed
}

# Trim whitespace
$old_ns1 =~ s/^\s+|\s+$//g;
$old_ns2 =~ s/^\s+|\s+$//g;
$new_ns1 =~ s/^\s+|\s+$//g;
$new_ns2 =~ s/^\s+|\s+$//g;

if ($old_ns1 eq $new_ns1 && $old_ns2 eq $new_ns2) {
    print "Error: Old and new namespaces are the same. Nothing to do.\n";
    exit(1);
}

print "Migrating from '$old_ns1::$old_ns2' to '$new_ns1::$new_ns2'...\n";
print "(Changing: $old_ns1 -> $new_ns1, $old_ns2 -> $new_ns2)\n\n";

# Track files modified
$files_modified = 0;

find(
    {
        wanted => \&findfiles,
    },
    'src'
);

find(
    {
        wanted => \&findfiles,
    },
    'tests'
);

print "\n=== Migration complete ===\n";
print "Total files modified: $files_modified\n";
print "\nRemember to:\n";
print "1. Review the changes with git diff\n";
print "2. Run cmake --build build --config Release\n\n";

sub findfiles
{
    $f = $File::Find::name;
    if ($f =~ m/\.h$/ or $f =~ m/\.cpp$/)
    {
        $q = basename($f);

        # Skip if we're already in the script's own processing (avoid recursion on .bak)
        return if ($q =~ /\.bak$/);

        # Read the entire file
        open(IN, "<$q") || die "Can't open IN $q: $!";
        @lines = <IN>;
        close(IN);

        $content = join("", @lines);
        $modified = 0;

        # Generic replacement 1: namespace oldns1 { -> namespace newns1 {
        if ($old_ns1 ne $new_ns1 && $content =~ s/namespace\s+$old_ns1\s*\{/namespace $new_ns1 {/g) {
            $modified = 1;
            print "Fixed namespace 1 opening in $q\n";
        }

        # Generic replacement 2: namespace oldns2 { -> namespace newns2 {
        if ($old_ns2 ne $new_ns2 && $content =~ s/namespace\s+$old_ns2\s*\{/namespace $new_ns2 {/g) {
            $modified = 1;
            print "Fixed namespace 2 opening in $q\n";
        }

        # Generic replacement 3: } // namespace oldns1 -> } // namespace newns1
        if ($old_ns1 ne $new_ns1 && $content =~ s/\}\s*\/\/\s*namespace\s+$old_ns1/} \/\/ namespace $new_ns1/g) {
            $modified = 1;
            print "Fixed namespace 1 closing comment in $q\n";
        }

        # Generic replacement 4: } // namespace oldns2 -> } // namespace newns2
        if ($old_ns2 ne $new_ns2 && $content =~ s/\}\s*\/\/\s*namespace\s+$old_ns2/} \/\/ namespace $new_ns2/g) {
            $modified = 1;
            print "Fixed namespace 2 closing comment in $q\n";
        }

        # Generic replacement 5: oldns1::oldns2:: -> newns1::newns2::
        if ($content =~ s/\Q$old_ns1\E::\Q$old_ns2\E::/$new_ns1::$new_ns2::/g) {
            $modified = 1;
            print "Fixed namespace references in $q\n";
        }

        # Generic replacement 6: namespace alias = oldns1::... -> namespace alias = newns1::...
        if ($old_ns1 ne $new_ns1 && $content =~ s/(namespace\s+\w+\s*=\s*)\Q$old_ns1\E::/${1}${new_ns1}::/g) {
            $modified = 1;
            print "Fixed namespace alias in $q\n";
        }

        # Critical fix: Ensure closing braces for multi-level namespaces are on separate lines
        # Look for: } // namespace ns3 } // namespace ns2 } // namespace ns1
        # and replace with proper multi-line format
        if ($content =~ s/\}\s*\/\/\s*namespace\s+(\w+)\s*\}\s*\/\/\s*namespace\s+\Q$old_ns2\E\s*\}\s*\/\/\s*namespace\s+\Q$old_ns1\E/} \/\/ namespace $1\n} \/\/ namespace $new_ns2\n} \/\/ namespace $new_ns1/g) {
            $modified = 1;
            print "Fixed multi-level namespace closing braces in $q\n";
        }

        # Update plugin ID if it contains the old namespace
        # Pattern: org.oldns1.oldns2 or similar
        if ($content =~ s/org\.\Q$old_ns1\E\.\Q$old_ns2\E/org.$new_ns1.$new_ns2/g) {
            $modified = 1;
            print "Fixed plugin ID in $q\n";
        }

        # Only write if modified
        if ($modified) {
            open(OUT, "> ${q}.bak") || die "Can't open BAK ${q}.bak: $!";
            print OUT $content;
            close(OUT);
            system("mv ${q}.bak $q");
            $files_modified++;
        }
    }
}

File: scripts/fix_namespace.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import re
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class NamespacePair:
    ns1: str
    ns2: str


def detect_current_namespaces(repo_root: Path) -> NamespacePair | None:
    files_to_check = [
        repo_root / "src" / "configuration.h",
        repo_root / "src" / "engine" / "engine.h",
        repo_root / "src" / "engine" / "patch.h",
        repo_root / "src" / "engine" / "voice.h",
    ]

    for file in files_to_check:
        if not file.exists():
            continue
        text = file.read_text(encoding="utf-8", errors="replace")

        m = re.search(r"\bnamespace\s+(\w+)::(\w+)(?:::.*)?\s*\{", text)
        if m:
            return NamespacePair(m.group(1), m.group(2))

        ns_stack: list[str] = []
        for line in text.splitlines():
            m2 = re.search(r"\bnamespace\s+(\w+)\s*\{", line)
            if m2:
                ns_stack.append(m2.group(1))
                if len(ns_stack) >= 2:
                    return NamespacePair(ns_stack[0], ns_stack[1])
            m3 = re.search(r"(\w+)::(\w+)::", line)
            if m3:
                return NamespacePair(m3.group(1), m3.group(2))

    return None


def apply_migration(text: str, old: NamespacePair, new: NamespacePair) -> tuple[str, bool]:
    original = text

    # Nested namespace openings: namespace old1::old2::sub {  -> namespace new1::new2::sub {
    text = re.sub(
        rf"\bnamespace\s+{re.escape(old.ns1)}::{re.escape(old.ns2)}((?:::\w+)*)\s*\{{",
        rf"namespace {new.ns1}::{new.ns2}\1 {{",
        text,
    )

    # Old style openings: namespace old1 { / namespace old2 {
    if old.ns1 != new.ns1:
        text = re.sub(rf"\bnamespace\s+{re.escape(old.ns1)}\s*\{{", f"namespace {new.ns1} {{", text)
    if old.ns2 != new.ns2:
        text = re.sub(rf"\bnamespace\s+{re.escape(old.ns2)}\s*\{{", f"namespace {new.ns2} {{", text)

    # Namespace references: old1::old2:: -> new1::new2::
    text = text.replace(f"{old.ns1}::{old.ns2}::", f"{new.ns1}::{new.ns2}::")

    # Namespace alias: namespace x = old1::... -> namespace x = new1::...
    if old.ns1 != new.ns1:
        text = re.sub(
            rf"(\bnamespace\s+\w+\s*=\s*){re.escape(old.ns1)}::",
            rf"\1{new.ns1}::",
            text,
        )

    # Closing comments for nested namespaces: } // namespace old1::old2...
    text = text.replace(
        f"// namespace {old.ns1}::{old.ns2}", f"// namespace {new.ns1}::{new.ns2}"
    )

    # Closing comments for old style: } // namespace old1 / old2
    if old.ns1 != new.ns1:
        text = re.sub(
            rf"(\}}\s*//\s*namespace\s+){re.escape(old.ns1)}\b", rf"\1{new.ns1}", text
        )
    if old.ns2 != new.ns2:
        text = re.sub(
            rf"(\}}\s*//\s*namespace\s+){re.escape(old.ns2)}\b", rf"\1{new.ns2}", text
        )

    # Split multi-close lines: } // namespace x } // namespace old2 } // namespace old1
    # (Best-effort; only handles this exact three-level inline pattern.)
    text = re.sub(
        rf"\}}\s*//\s*namespace\s+(\w+)\s*\}}\s*//\s*namespace\s+{re.escape(old.ns2)}\s*\}}\s*//\s*namespace\s+{re.escape(old.ns1)}",
        rf"}} // namespace \1\n}} // namespace {new.ns2}\n}} // namespace {new.ns1}",
        text,
    )

    # Plugin ID: org.old1.old2 -> org.new1.new2
    text = text.replace(f"org.{old.ns1}.{old.ns2}", f"org.{new.ns1}.{new.ns2}")

    return text, text != original


def iter_source_files(repo_root: Path) -> list[Path]:
    files: list[Path] = []
    for folder in ("src", "tests"):
        base = repo_root / folder
        if not base.exists():
            continue
        for ext in ("*.h", "*.cpp"):
            files.extend(base.rglob(ext))
    return files


def main() -> int:
    ap = argparse.ArgumentParser(description="Migrate the project's base namespace (two-level).")
    ap.add_argument("new_ns1", nargs="?", help="New first namespace (e.g., yourname)")
    ap.add_argument("new_ns2", nargs="?", help="New second namespace (e.g., yourproject)")
    ap.add_argument("--old-ns1", help="Override detected first namespace")
    ap.add_argument("--old-ns2", help="Override detected second namespace")
    ap.add_argument("--dry-run", action="store_true", help="Print what would change without writing files.")
    args = ap.parse_args()

    repo_root = Path(__file__).resolve().parent.parent

    detected = detect_current_namespaces(repo_root)
    if not detected and (not args.old_ns1 or not args.old_ns2):
        raise SystemExit("Could not auto-detect current namespaces. Provide --old-ns1 and --old-ns2.")

    old = NamespacePair(args.old_ns1 or detected.ns1, args.old_ns2 or detected.ns2)  # type: ignore[union-attr]

    if args.new_ns1 is None:
        raise SystemExit("Provide new_ns1 (and optionally new_ns2). Example: python scripts/fix_namespace.py myname myproject")

    new_ns2 = args.new_ns2 or old.ns2
    new = NamespacePair(args.new_ns1.strip(), new_ns2.strip())

    if old == new:
        raise SystemExit("Old and new namespaces are the same. Nothing to do.")

    print(f"Detected current namespaces: '{old.ns1}' :: '{old.ns2}'")
    print(f"Migrating to: '{new.ns1}' :: '{new.ns2}'")

    modified_files = 0
    for path in iter_source_files(repo_root):
        original = path.read_text(encoding="utf-8", errors="replace")
        updated, changed = apply_migration(original, old, new)
        if not changed:
            continue
        modified_files += 1
        if args.dry_run:
            print(f"Would update {path.relative_to(repo_root)}")
        else:
            path.write_text(updated, encoding="utf-8", newline="")
            print(f"Updated {path.relative_to(repo_root)}")

    print(f"Done. Files modified: {modified_files}{' (dry-run)' if args.dry_run else ''}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

